modules/database/src/std/rec/mbboDirectRecord.c

```diff
@@ -131,19 +139,7 @@ static long init_record(struct dbCommon *pcommon, int pass)
             status = 0;
     }
 
-    if (!prec->udf &&
-        prec->omsl == menuOmslsupervisory) {
-        /* Set initial B0 - B1F from VAL */
-        epicsUInt32 val = prec->val;
-        epicsUInt8 *pBn = &prec->b0;
-        int i;
-
-        for (i = 0; i < NUM_BITS; i++) {
-            *pBn++ = !! (val & 1);
-            val >>= 1;
-        }
-    }
-
+    bitsFromVAL(prec);
```

Comment:
Doesn't !UDF at this point imply that VAL has been initialized by `dbPutString()`?  (or explicitly cleared in a .db)

I'm looking for new cases for mbbioDirectTest.c.  What are the supported ways to initialize mbboDirect prior to a first `process()`?

```
record(mbboDirect, "A") {
    field(VAL, "3")
    field(B0, "1")
}
record(mbboDirect, "B") {
    field(B0, "1")
}
record(mbboDirect, "C") {
    field(VAL, "3")
}
```

Are there others?

Reply:
```
field(VAL, "3")
field(B0, "1")
```
is all well and good, but what is
```
field(VAL, "3")
field(B0, "0")
```
supposed to do? First set VAL to 3, then clear B0, resulting in VAL=2?
And what about
```
field(B0, "0")
field(VAL, "3")
```
Overwrite B0 with VAL=3? Having order dependent fields?

Reply:
Right... I had intended to show inconsistency.  Poor choice of example by me.

Ordering at least doesn't matter (I think) as the two fields are independent prior to `init_record()`

Reply:
The order of setting fields in a DB file can't matter (with the existing IOC code at least) because only record type code could act on fields outside of dbCommon, and `prec->rset` doesn't get set until iocInit.

Reply:
As all Bx and VAL default to 0, the only initialization that makes sense, in my opinion, is to OR all VAL and Bx fields.
```
field(VAL, "3")
field(B0, "0")
field(B4, "1")
```
Result: VAL=19 (0x13), B0,B1,B4=1

Reply:
@dirk-zimoch How do you intend for this should interact with autosave?  Remember that autosave pass0 happens before `init_record(prec, 0)`.

Reply:
Having VAL in autosave and Bx in the template will really cause confusion.
So what? Overwrite Bx with VAL but only if VAL!=0?

Reply:
> The order of setting fields in a DB file can't matter ...

wrt. `VAL` vs. `B*` this is true.  Excepting the special cases of `VAL` vs `UDF,` and `DTYP` vs. `INP/OUT` (and maybe others).

> ... template will really cause confusion.

Why yes, I believe it does.

> ... VAL!=0 ...

Looking at VAL alone doesn't give enough information.  eg. having non-zero B* by default, with autosave restoring zeros.

The one example I can find (of my own creation) is:

>    info(autosaveFields_pass0, "VAL")

So I guess the logic in `init_record()` should become:

```
if(!prec->udf) {
    bitsFromVAL(prec);
} else {
    // overwrite VAL from B*
}
```

Reply:
> `VAL` vs `UDF`

True.

> `DTYP` vs. `INP/OUT`

I believe you fixed that when you rewrote the address parser:
```
woz$ cat dtyp.db 
record(lso, lso1) {
  field(OUT, "@stdout")
  field(DTYP, "stdio")
}
record(lso, lso2) {
  field(DTYP, "stdio")
  field(OUT, "@stdout")
}
woz$ softIoc -d dtyp.db 
Starting iocInit
############################################################################
## EPICS R7.0.6.1-DEV
## Rev. R7.0.6-23-g21e7e4ddfbfbcf8ca5c6-dirty
############################################################################
iocRun: All initialization complete
epics> dbpr lso1 1
AMSG:               APST: On Change     ASG :               BKPT: 00            
DESC:               DISA: 0             DISP: 0             DISS: NO_ALARM      
DISV: 1             DOL : CONSTANT      DTYP: stdio         EVNT:               
FLNK: CONSTANT      LEN : 0             MPST: On Change     NAME: lso1          
NAMSG:              OMSL: supervisory   OUT : INST_IO @stdout                   
PACT: 0             PHAS: 0             PINI: NO            PRIO: LOW           
PUTF: 0             RPRO: 0             SCAN: Passive       SDIS: CONSTANT      
SEVR: INVALID       SIML: CONSTANT      SIMM: NO            SIOL: CONSTANT      
SIZV: 41            SSCN: <nil>         STAT: UDF           TPRO: 0             
TSE : 0             TSEL: CONSTANT      UDF : 1             UDFS: INVALID       
VAL : PTR 0x7fed6065fed0                
epics> dbpr lso2 1
AMSG:               APST: On Change     ASG :               BKPT: 00            
DESC:               DISA: 0             DISP: 0             DISS: NO_ALARM      
DISV: 1             DOL : CONSTANT      DTYP: stdio         EVNT:               
FLNK: CONSTANT      LEN : 0             MPST: On Change     NAME: lso2          
NAMSG:              OMSL: supervisory   OUT : INST_IO @stdout                   
PACT: 0             PHAS: 0             PINI: NO            PRIO: LOW           
PUTF: 0             RPRO: 0             SCAN: Passive       SDIS: CONSTANT      
SEVR: INVALID       SIML: CONSTANT      SIMM: NO            SIOL: CONSTANT      
SIZV: 41            SSCN: <nil>         STAT: UDF           TPRO: 0             
TSE : 0             TSEL: CONSTANT      UDF : 1             UDFS: INVALID       
VAL : PTR 0x7fed6065ffa0                
```

> Looking at VAL alone doesn't give enough information.

The contents of VAL can't tell you what to do. The original code in `init_record()` uses both UDF and OMSL to determine whether VAL or Bn should take precedence. If VAL was set either from a constant in DOL or by the device support's `init_record()` the UDF field will be FALSE and VAL takes precedence, so in `supervisory` mode it initializes Bn from VAL. Note that in this code `process()` constructs VAL from Bn when in `supervisory` mode.

In your new version the `process()` routine _always_ sets Bn from VAL, so your `init_record()` needs to initialize VAL from Bn when OMSL is `closed_loop` (unless UDF is FALSE, which indicates that a constant in DOL or the device support provided a value so it must then call `bitsFromVAL()`).

That's the change I was referring to in my original comment on this piece of code.